Add RMA to SM2

lrbison opened the pull request.

lrbison commented on prov/sm2/src/sm2.h:
```diff
@@ -143,6 +160,20 @@ static inline struct smr_freestack *sm2_freestack(struct sm2_region *smr)
+					uint64_t *device)
+{
+	const struct ofi_mr *mr = desc;
+	if (false && mr) {
+		*iface = mr->iface;
+		*device = mr->device;
+	} else {
+		*iface = FI_HMEM_SYSTEM;
+		*device = 0;
```
Fix me

lrbison commented on prov/sm2/src/sm2_attr.c:
```diff
@@ -104,7 +108,7 @@ struct fi_domain_attr sm2_domain_attr = {
 	.resource_mgmt = FI_RM_ENABLED,
 	.av_type = FI_AV_UNSPEC,
-	.mr_mode = FI_MR_BASIC | FI_MR_SCALABLE,
+	.mr_mode = OFI_MR_BASIC_MAP | FI_MR_LOCAL | FI_MR_BASIC,
```
Requesting feedback here to make sure these flags are right

shijin-aws replied:
It seems the current mode is copied from shm, anything wrong with it? May need to consult @aingerson

lrbison replied:
I switched them back and it sill works.  I don't remember why I changed this in the first place 🤷 .  Removing this change.

lrbison commented on prov/sm2/src/sm2_progress.c:
```diff
@@ -165,11 +167,257 @@ static int sm2_progress_recv_msg(struct sm2_ep *ep,
+	iov = cmd_rma->rma_iov;
+	payload_used = 0;
+	while (rma_count) {
+		// mr = (struct sm2_mr*)
+		// ofi_mr_map_get(&ep->base_ep.util_ep.domain->mr_map,
+		// (rta_hdr->rma_iov + i)->key);
```
Need to get HMEM working

shijin-aws commented on prov/sm2/src/sm2.h:
```diff
@@ -356,4 +356,14 @@ static inline size_t sm2_pop_xfer_entry(struct sm2_ep *ep,
+		container_of(ep->util_ep.av, struct sm2_av, util_av);
+	struct sm2_mmap *map = &av->mmap;
+	struct sm2_region *region = sm2_mmap_ep_region(map, ep->gid);
+	smr_freestack_push(sm2_freestack(region), xfer_entry);
```
Is `smr_freestack_push` in `ofi*.h`? The naming looks for shm only

lrbison replied:
yes, it is in ofi_mem.h.  The naming is... unfortunate to me.

a-szegel replied:
I have a backlog item in my notes to get us off of the SMR freestack... but yes, it should be renamed too.

shijin-aws commented on prov/sm2/src/sm2_rma.c:
```diff
@@ -0,0 +1,499 @@
+/*
+ * Copyright (c) 2013-2021 Intel Corporation. All rights reserved
```
Is it because we are copying it from shm?

lrbison replied:
Yes.  Originally this was smr_rma.c.  Much of it has changed, but not all of it.

shijin-aws reviewed (commented):
I reviewed with my limited knowledge. Didn't dive deep into the protocols

shijin-aws commented on prov/sm2/src/sm2.h:
```diff
@@ -335,14 +335,24 @@ static inline size_t sm2_pop_xfer_entry(struct sm2_ep *ep,
+		assert((char *) *xfer_entry >=
+		       (char *) sm2_freestack(self_region));
+		assert((char *) *xfer_entry <
+		       (char *) sm2_mmap_ep_region(map, ep->gid + 1));
```
So, `sm2_mmap_ep_region(map, ep->gid + 1)` is the beginning of the region for the next ep (if there is)? Does that mean the freestack is the only part of each region. This check does not look intuitive to me.

lrbison replied:
> So, `sm2_mmap_ep_region(map, ep->gid + 1)` is the beginning of the region for the next ep (if there is)?

Correct.  As I said above this is a crude check.  It's not a tight bound on the freestack but should be enough to find runaway problems.

I think it might actually be the last section of the region, but I'm not 100% sure.

lrbison replied:
Also, if you prefer I can just remove these asserts

shijin-aws replied:
No it doesn't hurt to check more things in DEBUG build, it would be good if you can add a comment here to explain

lrbison commented:
I've decided I want to change some names, also I find that we hang for 0-sized RMA.  Let me fix that.

a-szegel reviewed (changes requested):
first round of comments... didn't get to the main part yet.

a-szegel commented on prov/sm2/src/sm2_rma.c:
```diff
@@ -0,0 +1,509 @@
+				      comp_flags | FI_COMPLETION);
+		ret = FI_SUCCESS;
+		goto out;
+	}
```
This 0-byte read looks weird (comment would be nice), also it looks like it leaks a ctx object.

lrbison replied:
I have to remove it now that we claim RMA_EVENT capability.

a-szegel commented:
You need another commit where you enable the RMA fabtests (see Atomic PR for inspiration).

a-szegel commented on prov/sm2/src/sm2_rma.c:
```diff
@@ -0,0 +1,509 @@
+
+		src = (uint8_t *) msg->msg_iov[0].iov_base;
+		sm2_get_iface_device((void **) msg->desc, &iface, &device);
+		cmd_rma->rma_iov[jv].addr = msg->rma_iov[0].addr;
```
if we are always going to hard code rma_iov[0] to 0, should it even be an array?

The answer is that I think we need to send the full rma_iov so the receiving side knows where to put the data.

lrbison replied:
hinted above.  rma_iov[0] is always the right address for the remote to write into for this particular packet because we keep consuming the iov list which might fill the whole array.  Because we fill this rma list each time, the remote needs no context.

The alternative I think is to send the original, full RMA iov, then the remote needs to know which offset this packet corresponds to, does a "consume" of that many bytes, and uses the remaining values in the iov array to perform the copy of however many bytes are in the packet.  This also requires a packet size I suppose, since the RMA iov currently provides that for me.

a-szegel commented on prov/sm2/src/sm2_rma.c:
```diff
@@ -0,0 +1,509 @@
+	}
+	ctx->msgs_in_flight++;
+
+	return FI_SUCCESS;
```
for 0 byte operations, you are sending 1 message which I am guessing is for you to post a completion on the remote side. Do we want to do this?

lrbison replied:
yes, we must send a message in order for remote to post completion.  For 0-byte reads I avoided it, but for writedata we couldn't.  Now with RMA_EVENT capability, we must always send the message for 0-sized operations.  In fact, I had to re-order the loop to `do {...} while()` so that we send even for zero-sized operations.

a-szegel commented on prov/sm2/src/sm2_rma.c:
```diff
@@ -0,0 +1,509 @@
+		else {
+			sm2_rma_handle_local_error(ep, xfer_entry, rma_ctx,
+						   ret);
+			break;
```
It looks possible for you to have an error completion, and a transmit complete for messages (0 byte or messages that fail on last send).

a-szegel reviewed (changes requested):
still not done but here are more comments.

lrbison commented on prov/sm2/src/sm2.h:
```diff
@@ -88,6 +94,7 @@ extern pthread_mutex_t sm2_ep_list_lock;
 enum {
 	sm2_proto_inject,
 	sm2_proto_return,
+	sm2_proto_sar,
```
I expect it to be shared with message-based SAR, so I'm trying to leave it generic for now.

lrbison commented on prov/sm2/src/sm2.h:
```diff
@@ -129,6 +136,28 @@ struct sm2_ep_name {
+		uint32_t proto_flags;
+		size_t reserved;
+	};
+	void *context;
```
The generic header has the context that the user wants to see in completion.  This context is the state that the SAR protocol needs.  perhaps it could be called "sar_context" or "protocol_context"

lrbison replied:
done

lrbison commented on prov/sm2/src/sm2.h:
```diff
@@ -261,14 +261,27 @@ static inline size_t sm2_pop_xfer_entry(struct sm2_ep *ep,
 	}
-
 	*xfer_entry = smr_freestack_pop(sm2_freestack(self_region));
+#if ENABLE_DEBUG
```
ah, an older version had a variable and filled it prior to the asserts.  I can remove the block now.

lrbison commented on prov/sm2/src/sm2_attr.c:
```diff
@@ -33,17 +33,22 @@
-	(FI_SOURCE | OFI_RX_MSG_CAPS | FI_TAGGED | FI_DIRECTED_RECV | \
-	 FI_MULTI_RECV)
+#define SM2_TX_CAPS (OFI_TX_MSG_CAPS | FI_TAGGED | OFI_TX_RMA_CAPS)
+#define SM2_RX_CAPS                                                  \
```
Added to RX capabilities.

lrbison replied:
Removed it again.  There is some completion issue during testing.

lrbison commented on prov/sm2/src/sm2_attr.c:
```diff
@@ -154,7 +159,8 @@ struct fi_info sm2_hmem_info = {
 struct fi_info sm2_info = {
-	.caps = SM2_TX_CAPS | SM2_RX_CAPS | FI_MULTI_RECV | FI_LOCAL_COMM,
+	.caps = SM2_TX_CAPS | SM2_RX_CAPS | FI_MULTI_RECV | FI_LOCAL_COMM |
+		SM2_RMA_CAPS,
```
done

lrbison commented on prov/sm2/src/sm2_attr.c:
```diff
@@ -33,17 +33,22 @@
 	 FI_DELIVERY_COMPLETE)
 #define SM2_RX_OP_FLAGS (FI_COMPLETION | FI_MULTI_RECV)
 
+#define SM2_RMA_CAPS                                                      \
```
done

lrbison replied:
Oops, this snuck back in during modifications.  The define is present but the value is unused.

lrbison commented on prov/sm2/src/sm2_ep.c:
```diff
@@ -225,7 +227,8 @@ static void cleanup_shm_resources(struct sm2_ep *ep)
 return_incoming:
 	while (NULL != (xfer_entry = sm2_fifo_read(ep))) {
-		if (xfer_entry->hdr.proto == sm2_proto_return) {
+		if (xfer_entry->hdr.proto == sm2_proto_return ||
```
Yes.  Perhaps we should rename it from "sender_gid" to "owner_gid" as well.

lrbison replied:
we can address the `s/sender_gid/owner_gid` this in the future

lrbison commented on prov/sm2/src/sm2_ep.c:
```diff
@@ -564,13 +570,20 @@ int sm2_endpoint(struct fid_domain *domain, struct fi_info *info,
+	}
+	ret = ofi_bufpool_create(&ep->sar_ctx_pool, sizeof(struct sm2_sar_ctx),
+				 16, 0, info->rx_attr->size,
+				 OFI_BUFPOOL_NO_TRACK);
```
Why?  What fixed size would you prefer?

a-szegel replied:
`SM2_NUM_XFER_ENTRY_PER_PEER` which I believe will one day become tx_size

lrbison replied:
done

lrbison reviewed (commented):
Thanks for the feedback Seth.  I've addressed your changes so far in my local copy.  I'll save the CI server a few cycles and wait to push until you've finished reviewing.

lrbison commented on prov/sm2/src/sm2.h:
```diff
@@ -261,14 +261,27 @@ static inline size_t sm2_pop_xfer_entry(struct sm2_ep *ep,
 		container_of(ep->util_ep.av, struct sm2_av, util_av);
 	struct sm2_mmap *map = &av->mmap;
 	struct sm2_region *self_region = sm2_mmap_ep_region(map, ep->gid);
+	*xfer_entry = NULL;
```
done

lrbison commented on prov/sm2/src/sm2.h:
```diff
@@ -78,6 +78,12 @@
 #define SM2_IOV_LIMIT		4
 #define SM2_INJECT_SIZE		(SM2_XFER_ENTRY_SIZE - sizeof(struct sm2_xfer_hdr))
 
+#define FI_SM2_SAR_FIRST_MESSAGE_FLAG (1 << 0)
```
removed

lrbison commented on prov/sm2/src/sm2_rma.c:
```diff
@@ -0,0 +1,509 @@
+			if (rma_ctx->msgs_in_flight == 0) {
+				sm2_free_rma_ctx(rma_ctx);
+				ret = -FI_EAGAIN;
+			} else
```
Agree, just trying to follow the rules but I was informed of this caveat last week!

lrbison commented on prov/sm2/src/sm2_rma.c:
```diff
@@ -0,0 +1,509 @@
+		if (ctx->bytes_sent == ctx->bytes_total)
+			break;
+		bytes_to_send =
+			MIN(msg->msg_iov[0].iov_len, msg->rma_iov[0].len);
```
I'll rework this.  I think you are right that it's not quite correct.  In cases where there are non-overlapping remote and local iovs we may have empty space left in our packets.

The zero-indexing is intentional though, because we are calling `ofi_consume_*` which will eat some of the iov and leave `msg_iov[0]` as the next data to send.

lrbison replied:
resolved.

a-szegel commented on prov/sm2/src/sm2_progress.c:
```diff
@@ -202,6 +203,278 @@ static int sm2_progress_recv_msg(struct sm2_ep *ep,
+		return FI_SUCCESS;
+	}
+
+	if (ctx->bytes_requested == ctx->bytes_total) {
```
When do we get into this case... are we leaking a ctx?

lrbison replied:
No, this is similar to rx_complete/tx_complete but for read.  Essentially we don't need to send back the packet requesting more data, because we already have sent enough requests, we just need to return this entry to freestack and wait for the last request to return to us.

a-szegel commented on prov/sm2/src/sm2_progress.c:
```diff
@@ -202,6 +203,278 @@ static int sm2_progress_recv_msg(struct sm2_ep *ep,
+	bool tx_complete, rx_complete;
+	int ret;
+
+	if (!ctx) {
```
It looks to me like rma_inject useses generic_rma which creates a context for every message... so this branch is dead code which will never be hit.

lrbison replied:
Thank you, old code :-)

a-szegel commented on prov/sm2/src/sm2_progress.c:
```diff
@@ -41,6 +41,7 @@
 #include "ofi_mr.h"
 #include "sm2.h"
 #include "sm2_fifo.h"
+#include "sm2_rma.h"
```
I think we should put these functions in sm2_rma.c ... what do you think?

lrbison replied:
I wasn't sure which functions you meant here.  You mean alloc and free?  We still need free in progress as well.

a-szegel commented on prov/sm2/src/sm2_rma.h:
```diff
@@ -0,0 +1,52 @@
+	ofi_buf_free(rma_ctx);
+}
+
+void sm2_fill_rma_ctx(struct sm2_ep *ep, const struct fi_msg_rma *msg,
```
nit: i sort of hate declarations between definitions, can we have all the declarations above or below the definitions?

lrbison replied:
Resolved when I moved some functions into sm2_fifo.h

a-szegel reviewed (changes requested):
Finished my review, very well done overall.  Thank you Luke!

lrbison commented on prov/sm2/src/sm2_progress.c:
```diff
@@ -202,6 +203,278 @@ static int sm2_progress_recv_msg(struct sm2_ep *ep,
+		src += iov->len;
+		assert((long) (src - (uint8_t *) &cmd_rma->user_data) <=
+		       SM2_RMA_INJECT_SIZE);
+		if (ret) {
```
Addressed.  Sequence for this case is:
1. Remote (the ep in this progress function) emits error to CQ
2. Origin emits error to CQ
3. Origin cleans up context.

As long as remote has no context and only 1 message is in flight at a time, this is sufficient.

If we want more than 1 message in flight at a time, we need some state at remote to prevent a subsequent message from  succeeding and accidentally submitting 2 CQs.

lrbison commented on prov/sm2/src/sm2_progress.c:
```diff
@@ -202,6 +203,278 @@ static int sm2_progress_recv_msg(struct sm2_ep *ep,
+		}
+	}
+
+	if (cmd_rma->sar_hdr.proto_flags & FI_SM2_SAR_ERROR_FLAG)
```
fixed using goto out rather than break in the loop, and out skips the path that might generate successful completion.

lrbison commented on prov/sm2/src/sm2_progress.c:
```diff
@@ -202,6 +203,278 @@ static int sm2_progress_recv_msg(struct sm2_ep *ep,
 	return ret < 0 ? ret : 0;
 }
 
+static inline void sm2_rma_write_back(struct sm2_ep *ep,
```
moved to sm2_fifo.

lrbison commented on prov/sm2/src/sm2_progress.c:
```diff
@@ -202,6 +203,278 @@ static int sm2_progress_recv_msg(struct sm2_ep *ep,
+
+	while (iov_count) {
+		bytes_to_copy =
+			MIN(iov[0].iov_len, SM2_RMA_INJECT_SIZE - bytes_used);
```
consume does the work for us.

lrbison replied:
latest update does this now.

lrbison commented:
I think I've addressed most if not all comments.  I did add ubertests, but there are some tests that it complains about ENODATA.  We'll see what CI says about that.  They did help my find one additional problem: I had the wrong level of pointer indirection for `desc` which made multiple iov RMA operations not work.

One additional problem I discovered is that when I added FI_RMA_EVENT capabilities, I have all sorts of problems with the fi_rma_bw fabtest.  SHM seems to have this capability and passes the test fine, so there must be something wrong in my implementation.  It seemed that the completions I was generating for writes on remote was being interpreted as message completions and so it would not sync correctly.  Oddly, I'd even get verification errors on the message ft_sync() itself was sending (probably because the message never arrived but the CQ counter incremented because of RMA).

For now I've removed FI_RMA_EVENT capabilities and this solves the immediate problem.

vidsouza commented:
bot:aws:retest

a-szegel commented:
bot:aws:retest

a-szegel commented:
closing to https://github.com/ofiwg/libfabric/pull/9089

a-szegel reviewed (approved):
Looks good!  Thank you :)

aingerson commented on prov/sm2/src/sm2.h:
```diff
@@ -289,6 +289,9 @@ static inline size_t sm2_pop_xfer_entry(struct sm2_ep *ep,
 	*xfer_entry = smr_freestack_pop(sm2_freestack(ep->self_region));
+
+	/* ensure we haven't over-exhausted the freestack: */
+	assert(sm2_freestack(ep->self_region)->free >= 0);
```
I think this assert would make more sense before you pop
`assert(sm2_freestack(ep->region)->free);`

lrbison replied:
I removed some other checks this commit originally had.  I think now this commit is not so useful.  I'll remove it.

aingerson commented on prov/sm2/src/sm2_fifo.h:
```diff
@@ -241,4 +241,30 @@ static inline void sm2_fifo_write_back(struct sm2_ep *ep,
+	static char outstr[80];
+	struct sm2_av *av;
+	struct sm2_mmap *map;
+	av = container_of(ep->util_ep.av, struct sm2_av, util_av);
+	map = &av->mmap;
+	struct sm2_coord_file_header *header = (void *) map->base;
```
Fix variable declaration block - declare (inline initialization is ok), space, rest of code.
Also I think Seth added a helper function to simplify the map pointer get

aingerson commented on prov/sm2/src/sm2_fifo.h:
```diff
@@ -241,4 +241,30 @@ static inline void sm2_fifo_write_back(struct sm2_ep *ep,
+	av = container_of(ep->util_ep.av, struct sm2_av, util_av);
+	map = &av->mmap;
+	struct sm2_coord_file_header *header = (void *) map->base;
+	if (map->base > (char *) ptr)
+		sprintf(outstr, "%p (not a mapped pointer, address before map)",
+			ptr);
+	else if (map->base + map->size < (char *) ptr)
+		sprintf(outstr, "%p (not a mapped pointer, address after map)",
+			ptr);
+	else {
```
Fix bracket use

lrbison replied:
fixed

aingerson commented on prov/sm2/src/sm2_fifo.h:
```diff
@@ -241,4 +241,30 @@ static inline void sm2_fifo_write_back(struct sm2_ep *ep,
+		sprintf(outstr, "%p (%ld, %ld bytes into region %d)", ptr,
+			relptr, region_off, jregion);
+	}
+	return outstr;
```
Returned local variable isn't accessible after the function exits. See what the fi_tostr function does to get around it.

lrbison replied:
The variable is static, so it should be allocated in the data segment.  I suppose we are uselessly pre-allocating it even though we don't expect it to be used (this is primarily for debug support convenience).  I'll make it a NULL pointer and allocate it if needed.

lrbison replied:
on second thought that might make valgrind mad.  Let's just leave it as is.

aingerson commented on prov/sm2/src/sm2_fifo.h:
```diff
@@ -241,4 +241,30 @@ static inline void sm2_fifo_write_back(struct sm2_ep *ep,
 	sm2_fifo_write(ep, xfer_entry->hdr.sender_gid, xfer_entry);
 }
 
+static inline const char *sm2_print_ptr(struct sm2_ep *ep, void *ptr)
```
Please add more of a description in the commit explaining what exactly this does (who owns the ptr, what look up is it doing, etc)

aingerson replied:
That's fine but please put it in both places

lrbison replied:
done

aingerson commented on prov/sm2/src/sm2.h:
```diff
@@ -299,4 +299,10 @@ int sm2_query_atomic(struct fid_domain *domain, enum fi_datatype datatype,
+inline static void sm2_freestack_push(struct sm2_ep *ep,
+				      struct sm2_xfer_entry *xfer_entry)
+{
+	smr_freestack_push(sm2_freestack(ep->self_region), xfer_entry);
```
This seems like a rather silly function to have... it's pretty clear what it's doing and is only called twice. what is the purpose of this?

lrbison replied:
well, when I wrote it there were locks involved and ep->self_region were not present.  It was originally much longer to account for that.  Also although it's only used twice during this commit, it will get used a total of 8 times in later commits.

Finally, when scrolling through code with a bunch of `sm2_*` functions, I find it useful to have push be an `sm2_*` function.

aingerson commented on prov/sm2/src/sm2_rma.h:
```diff
@@ -0,0 +1,34 @@
+#include "sm2.h"
```
Missing copyright.
Also I don't think this separate file is really necessary? All of these can just be static inside rma.c/progress.c

lrbison replied:
ack on copyright thanks.

I actually would rather have all of these be in a sm2_sar.c/h file and have progress jump into various protocols, but I don't want to refactor current organization right now.

Would you prefer this file be named sm2_sar.h?

aingerson replied:
I'm ok with having a separate header but it really needs to be isolated to what it does (whether SAR or RMA). It looks like this does both which is why it's confusing.

lrbison replied:
Removed the rma.h header and renamed several of those functions to be sar specific rather than rma specific.  You can find them in sm2.h now.

aingerson commented on prov/sm2/src/sm2.h:
```diff
@@ -82,6 +82,15 @@
 #define SM2_ATOMIC_COMP_INJECT_SIZE (SM2_ATOMIC_INJECT_SIZE / 2)
 
+/* protocol flags */
+#define SM2_SAR_LAST_MESSAGE_FLAG (1 << 1)
+#define SM2_SAR_RETURN		  (1 << 2)
+#define SM2_SAR_ERROR_FLAG	  (1 << 3)
+#define SM2_SAR_STATUS_COMPLETED  (1 << 4)
+
```
This patch is adding RMA but it looks like it's adding SAR stuff as well? Please separate

lrbison replied:
We intend on introducing SAR for messages and for RMA.  Rather than have one massive PR I've started with implementing RMA using SAR, and intend to do messages with SAR next.  One of them has to come first.  I cannot have a commit introducing SAR without having some mechanism which uses it.

aingerson replied:
I think it's very confusing to have it come in as one. SAR and RMA are two completely different things. SAR is a protocol. RMA is API functionality. SAR will be needed for messages and RMA (the flow should be totally the same with both, just the way you resolve the destination address is different) so doesn't rely on RMA. RMA should also work the same as the other protocols as well so you should be able to use RMA with inline/inject so it shouldn't rely on SAR. SAR and RMA are completely separate. I think avoiding massive PRs is great and I think that's what this is if you combine them. 
I think you need to either add SAR for regular messages first and then add RMA for all protocols or add RMA for the existing protocols and then add SAR for both. Either order is fine with me, but merging them into one is extremely confusing and hard to read and review.

lrbison replied:
> SAR is a protocol. RMA is API functionality.

100% Agree and I understand what you are saying.  However I am not a prescient developer.  I have a concept for a SAR protocol in my head, and I have a concept of how that protocol fits into implementing RMA and messaging functionalities.

It seemed like RMA capability was a good place to start because other developers were providing large message capabilities, and RMA was the sore-thumb of missing functionality.

I did my best to clearly delineate SAR that I expected to re-use in the future for messaging by including SAR in the name, however as I made progress my mental model of which functions would be re-used and how was adjusted, and I'm sure they'll continue to adjust when messages are implemented.  Please don't mistake SAR being used for many RMA operations in this commit to mean that I have somehow confused the two.

I do expect that the Messaging-using-SAR commit will probably further solidify and clarify where the boundaries fall, but given the pace of development, the number of revisions each PR gets, and the amount of parallel work which would be hurt by merge conflicts, I think the choice to implement RMA with SAR and then implement Messages with SAR is the best way forward.  These both need to happen because the SAR-protocol path provides large message when CMA is not enabled and this PR should provide all-size support for RMA, even if in the future we have a faster option.

> you should be able to use RMA with inline/inject so it shouldn't rely on SAR.
The SAR I've written gracefully degrades to something akin to inject, but I'm using the SAR header, so it remains "SAR".

aingerson replied:
Sorry, I wasn't suggesting you were confusing the two. I was saying that putting them together in the same patch is confusing to me as a reviewer because it's hard to tell if a change has to do with SAR or RMA specifically. 
I understand that simultaneous development causes headaches but I think it's for that reason to split it out into easier to handle chunks. Taking the easy way out of writing a lot of code and then just merging it to avoid conflicts so people can rebase I think is problematic for the project as a whole. If simultaneous development within your team is causing issues because you are relying on all the code to go into upstream in order for others to pick it up, I would suggest using someone's branch as a development playground instead (ie have everyone merge to Seth's AWS_sm2 branch and everyone can rebase off that) and then squash commits accordingly. There's no reason to rush things into upstream that aren't fully thought out.
If the SAR you have written gracefully degrades to inject, I would think that it would be easy to plug the RMA you've written into the existing inject protocol. That change, I think, would be simple and could be merged right away. My concern with them together is that SAR is a much more complicated beast since it requires multiple round trips and that requires more thorough review. I would rather remove the RMA specific parts of the code so that we can better view the SAR protocol to understand how it moves data around, whether it's a send or a read/write.

lrbison replied:
I'll do my best to split RMA into a commit with inject only followed by SAR.

aingerson reviewed (commented):
I'm going to submit my initial review early (still going to finish, but it needs some larger adjustments so I'll let you get started on them).
The RMA patch looks like it's adding RMA as well as SAR so that needs to be 2 separate commits.
Also please get in the habit of always including information in the commit message. Some are ok empty (small change, one line, etc). But for the most part, more information is better. The RMA patch especially (and subsequent SAR patch as well) needs more information explaining the flow of the protocols. It makes it much easier for reviewing and for future developers to look at commits and understand what they did/are doing.

sunkuamzn reviewed (commented):
Just a couple of minor comments. Thanks Luke!

lrbison commented on prov/sm2/src/sm2_attr.c:
```diff
@@ -86,14 +89,15 @@ struct fi_rx_attr sm2_hmem_rx_attr = {
 	.iov_limit = SM2_IOV_LIMIT,
 };
 
+#define SMALLEST_INJECT MIN(SM2_RMA_INJECT_SIZE, SM2_ATOMIC_INJECT_SIZE)
```
This define is really a line-length convenience.  It's only scope is in this .c file.  I can add SM2_ at the front.

lrbison replied:
done

lrbison commented on prov/sm2/src/sm2_progress.c:
```diff
@@ -330,6 +458,138 @@ static int sm2_progress_atomic(struct sm2_ep *ep,
+		comp_flags = ofi_rx_cq_flags(xfer_entry->hdr.op);
+		comp_flags |= FI_REMOTE_CQ_DATA & xfer_entry->hdr.op_flags;
+		/* TODO: when sm2 supports FI_RMA_EVENT: */
+		/* comp_flags |= FI_COMPLETION & xfer_entry->hdr.op_flags; */
```
I suppose I could remove both comments and let whoever implements sm2's FI_RMA_EVENT re-discover this, or I could leave these breadcrumbs.

I guess I'll remove.

lrbison commented:
Closing this as OBE.
